eventstore/manager.go

```diff
@@ -0,0 +1,114 @@
+package eventstore
+
+import (
+	"io"
+
+	"github.com/google/uuid"
+	ds "github.com/ipfs/go-datastore"
+	kt "github.com/ipfs/go-datastore/keytransform"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/textileio/go-textile-core/threadservice"
+)
+
+var (
+	dsStoreManagerBaseKey = ds.NewKey("/manager")
+)
+
+type Manager struct {
+	io.Closer
+
+	config *StoreConfig
+
+	threadservice threadservice.Threadservice
+	stores        map[uuid.UUID]*Store
+}
+
+// NewManager hydrates stores from prefixes and start them.
+func NewManager(ts threadservice.Threadservice, opts ...StoreOption) (*Manager, error) {
+	config := &StoreConfig{}
+	for _, opt := range opts {
+		if err := opt(config); err != nil {
+			return nil, err
+		}
+	}
+
+	if config.Datastore == nil {
+		datastore, err := newDefaultDatastore(config.RepoPath)
+		if err != nil {
+			return nil, err
+		}
+		config.Datastore = datastore
+	}
+
+	// @todo: handle debug here or in store?
+
+	m := &Manager{
+		config:        config,
+		threadservice: ts,
+		stores:        make(map[uuid.UUID]*Store),
+	}
+
+	q, err := m.config.Datastore.Query(query.Query{
+		Prefix:   dsStoreManagerBaseKey.String(),
+		KeysOnly: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer q.Close()
+
+	for res := range q.Next() {
+		idStr := ds.RawKey(res.Key).Parent().Parent().Parent().Parent().Name() // reaching for the stars here
+		if idStr == "" {
+			continue
+		}
+		id, err := uuid.Parse(idStr)
+		if err != nil {
+			return nil, err
+		}
+		s, err := newStore(m.threadservice, m.config)
+		if err != nil {
+			return nil, err
+		}
+		s.datastore = kt.Wrap(s.datastore, kt.PrefixTransform{
+			Prefix: dsStoreManagerBaseKey.ChildString(id.String()),
+		})
+		if err = s.Start(); err != nil {
+			return nil, err
+		}
+		m.stores[id] = s
+	}
+
+	return m, nil
+}
+
+// NewStore creates a new store and prefix its datastore with base key.
+func (m *Manager) NewStore() (id uuid.UUID, store *Store, err error) {
+	id, err = uuid.NewRandom()
+	if err != nil {
+		return
+	}
+	store, err = newStore(m.threadservice, m.config)
```

Comment:
All managed stores are created with the same config, which contains the same datastore and other settings.